zk/dnshostprovider_test.go

```diff
@@ -165,6 +166,47 @@ func TestDNSHostProviderReconnect(t *testing.T) {
 	}
 }
 
+// TestDNSHostOneHostDead tests whether
+func TestDNSHostOneHostDead(t *testing.T) {
+	hp := &DNSHostProvider{lookupHost: func(host string) ([]string, error) {
+		if host == "foo.failure.com" {
+			return nil, errors.New("Fails to ns lookup")
+		}
+		return []string{"192.0.2.1", "192.0.2.2"}, nil
+	}, sleep: func(_ time.Duration) {}}
+
+	if err := hp.Init([]string{"foo.failure.com:12345", "foo.success.com:12345"}); err != nil {
+		t.Fatal(err)
+	}
+
+	hp.mu.Lock()
+	if len(hp.servers) != 2 {
+		t.Fatal("Only servers that resolved by lookupHost should be in servers list")
+	}
+
+	// update lookupHost to mock a successful lookup
+	hp.lookupHost = func(host string) ([]string, error) {
+		if host == "foo.failure.com" {
+			return []string{"192.0.2.3"}, nil
+		}
+		return []string{"192.0.2.1", "192.0.2.2"}, nil
+	}
+	hp.mu.Unlock()
+
+	// Starts a 30s retry loop to wait servers list to be updated
+	startRetryLoop := time.Now()
+	for {
+		hp.mu.Lock()
```

Comment:
I'd add back the sleep as the first step of this for loop, to avoid pegging the CPU. sleeps *are* useful in tests like these, they just need to be wrapped in a loop like this to avoid CI flakes :)

Reply:
Good idea. Fixed.